src/display/editor/outliner.js

```diff
@@ -331,4 +343,475 @@ class HighlightOutline extends Outline {
+  }
+
+  serialize([blX, blY, trX, trY], rotation) {
+    const src = this.#outline;
+    const outline = new Float64Array(src.length);
+    const width = trX - blX;
+    const height = trY - blY;
+    switch (rotation) {
+      case 0:
+        for (let i = 0, ii = src.length; i < ii; i += 2) {
+          outline[i] = blX + src[i] * width;
+          outline[i + 1] = trY - src[i + 1] * height;
+        }
+        break;
+      case 90:
+        for (let i = 0, ii = src.length; i < ii; i += 2) {
+          outline[i] = blX + src[i + 1] * width;
+          outline[i + 1] = blY + src[i] * height;
+        }
+        break;
+      case 180:
+        for (let i = 0, ii = src.length; i < ii; i += 2) {
+          outline[i] = trX - src[i] * width;
+          outline[i + 1] = blY + src[i + 1] * height;
+        }
+        break;
+      case 270:
+        for (let i = 0, ii = src.length; i < ii; i += 2) {
+          outline[i] = trX - src[i + 1] * width;
+          outline[i + 1] = trY - src[i] * height;
+        }
+    }
+    return outline;
+  }
+
+  #computeMinMax(isLTR) {
+    const outline = this.#outline;
+    let lastX = outline[4];
+    let lastY = outline[5];
+    let minX = lastX;
+    let minY = lastY;
+    let maxX = lastX;
+    let maxY = lastY;
+    let lastPointX = lastX;
+    let lastPointY = lastY;
+    const ltrCallback = isLTR ? Math.max : Math.min;
+
+    for (let i = 6, ii = outline.length; i < ii; i += 6) {
+      if (isNaN(outline[i])) {
+        minX = Math.min(minX, outline[i + 4]);
+        minY = Math.min(minY, outline[i + 5]);
+        maxX = Math.max(maxX, outline[i + 4]);
+        maxY = Math.max(maxY, outline[i + 5]);
+        if (lastPointY < outline[i + 5]) {
+          lastPointX = outline[i + 4];
+          lastPointY = outline[i + 5];
+        } else if (lastPointY === outline[i + 5]) {
+          lastPointX = ltrCallback(lastPointX, outline[i + 4]);
+        }
+      } else {
+        const bbox = Util.bezierBoundingBox(
+          lastX,
+          lastY,
+          ...outline.slice(i, i + 6)
+        );
+        minX = Math.min(minX, bbox[0]);
+        minY = Math.min(minY, bbox[1]);
+        maxX = Math.max(maxX, bbox[2]);
+        maxY = Math.max(maxY, bbox[3]);
+        if (lastPointY < bbox[3]) {
+          lastPointX = bbox[2];
+          lastPointY = bbox[3];
+        } else if (lastPointY === bbox[3]) {
+          lastPointX = ltrCallback(lastPointX, bbox[2]);
+        }
+      }
+      lastX = outline[i + 4];
+      lastY = outline[i + 5];
+    }
+
+    this.#lastPoint = [lastPointX, lastPointY];
+    this.#bbox = {
+      x: minX - this.#innerMargin,
+      y: minY - this.#innerMargin,
+      width: maxX - minX + 2 * this.#innerMargin,
+      height: maxY - minY + 2 * this.#innerMargin,
+      lastPoint: this.#lastPoint,
+    };
+    this.#lastPoint[0] = (this.#lastPoint[0] - this.#bbox.x) / this.#bbox.width;
+    this.#lastPoint[1] =
+      (this.#lastPoint[1] - this.#bbox.y) / this.#bbox.height;
```

Comment:
Is the `this.#lastPoint` field still necessary now, since this seems to be the only spot in this class where it's used now?
Could we perhaps do something like the following instead?
```suggestion
    const x = minX - this.#innerMargin,
      y = minY - this.#innerMargin,
      width: maxX - minX + 2 * this.#innerMargin,
      height: maxY - minY + 2 * this.#innerMargin;
      lastPoint: [(lastPointX - x) / width, (lastPointY - y) / height];
    this.#bbox = { x, y, width, height, lastPoint, };
```